structure_tests/drivers/dockerDriver.go

```diff
@@ -0,0 +1,298 @@
+	"github.com/fsouza/go-dockerclient"
+)
+
+type DockerDriver struct {
+	OriginalImage string
+	CurrentImage  string
+	cli           docker.Client
+}
+
+func (d DockerDriver) Info() string {
+	return fmt.Sprintf("DockerDriver:\nOriginalImage: %s\nCurrentImage: %s\ncli: %s\n",
+		d.OriginalImage, d.CurrentImage, d.cli)
+}
+
+func NewDockerDriver(image string) DockerDriver {
+	newCli, err := docker.NewClientFromEnv()
+	if err != nil {
+		panic(err)
+	}
+	return DockerDriver{
+		OriginalImage: image,
+		CurrentImage:  image,
+		cli:           *newCli,
+	}
+}
+
+func (d DockerDriver) ProcessCommand(t *testing.T, envVars []unversioned.EnvVar, fullCommand []string,
+	shellMode bool, checkOutput bool) (string, string, int) {
+
+	if len(fullCommand) == 0 {
+		t.Logf("empty command provided: skipping...")
+		return "", "", -1
+	}
+	env := []string{}
+	if len(envVars) > 0 {
+		for _, envVar := range envVars {
+			env = append(env, envVar.Key+"="+envVar.Value)
+		}
+	}
+	stdout, stderr, exitCode := d.exec(t, env, fullCommand)
+
+	if stdout != "" {
+		t.Logf("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		t.Logf("stderr: %s", stderr)
+	}
+	//reset image for next test
+	d.CurrentImage = d.OriginalImage
+	return stdout, stderr, exitCode
+}
+
+func (d DockerDriver) SetEnvVars(t *testing.T, vars []unversioned.EnvVar) []unversioned.EnvVar {
+	// currently, this does not support substitution, so setting env vars is broken
+	// need to find a way to retrieve the existing values from the container before setting
+	if len(vars) == 0 {
+		return nil
+	}
+	env := []string{}
+	for _, envVar := range vars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	newImage := d.runAndCommit(t, env, nil)
+	// since these are global envvars, just overwrite the original image
+	d.OriginalImage = newImage
+	d.CurrentImage = d.OriginalImage
+	return nil
+}
+
+func (d DockerDriver) ResetEnvVars(t *testing.T, vars []unversioned.EnvVar) {
+	// since the container will be destroyed after the tests, this is a noop
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d DockerDriver) retrieveFile(t *testing.T, path string, directory bool) (string, error) {
+	ctx := context.Background()
+
+	// TODO(nkubala): this contains a hack to get around the fact that
+	// the client will not allow creating a container without a command.
+	// we should remove this, or at the very least find a better alternative
+	// given that not every container is guaranteed to have a shell.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.CurrentImage,
+			Cmd:          []string{"/bin/sh"},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+		Context:          ctx,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", err
+	}
+
+	if directory {
```

Comment:
Why would this ever be called with directory true?

Reply:
ah that's an option in the structure test config, you can specify whether something is a directory if you want to check that instead. this still needs to be implemented